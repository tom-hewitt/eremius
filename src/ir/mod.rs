mod tests;

#[derive(Debug, PartialEq)]
pub enum InstructionKind<
    BranchAddress = SignedImm<24>,
    LoadStoreAddress = LoadStoreAddressCode<Imm<12>, Imm<5>>,
    ShifterOperand = ShifterOperandCode<RotatedImm8, Imm<5>>,
    SuperVisorCallNumber = Imm<24>,
> {
    Branch {
        condition: Condition,
        kind: BranchKind,
        target: BranchAddress,
    },

    DataProcessing {
        condition: Condition,
        kind: DataProcessingKind<ShifterOperand>,
    },

    LoadStore {
        condition: Condition,
        kind: LoadStoreKind,
        quantity: LoadStoreQuantity,
        destination: Rd,
        address: LoadStoreAddress,
    },
    LoadStoreMultiple {
        condition: Condition,
        kind: LoadStoreKind,
        mode: MultipleAddressingMode,
        base: Rn,
        write_back: WriteBack,
        register_list: RegisterList,
    },

    SuperVisorCall {
        condition: Condition,
        immediate: SuperVisorCallNumber,
    },
}

#[derive(Debug, PartialEq)]
pub enum DataProcessingKind<ShifterOperand = ShifterOperandCode> {
    Move {
        kind: MoveKind,
        set_flags: SetFlags,
        destination: Rd,
        shifter: ShifterOperand,
    },
    Comparison {
        kind: ComparisonKind,
        source: Rn,
        shifter: ShifterOperand,
    },
    Calculation {
        kind: CalculationKind,
        set_flags: SetFlags,
        destination: Rd,
        source: Rn,
        shifter: ShifterOperand,
    },
}

#[derive(Debug, PartialEq)]
pub enum ComparisonKind {
    CMP,
    // CMN,
    // TST,
    // TEQ,
}

#[derive(Debug, PartialEq)]
pub enum CalculationKind {
    ADD,
    SUB,
    // RSB,
    // ADC,
    // SBC,
    // RSC,
    // AND,
    // BIC,
    // EOR,
    // ORR,
}

#[derive(Debug, PartialEq)]
pub enum LoadStoreKind {
    Load,
    Store,
}

#[derive(Debug, PartialEq)]
pub struct LoadStoreAddressCode<Immediate, ShiftImmediate> {
    pub base: Rn,
    pub offset: AddressingOffset<Immediate, ShiftImmediate>,
}

#[derive(Debug, PartialEq)]
pub enum Sign {
    Positive,
    Negative,
}

#[derive(Debug, PartialEq)]
pub struct AddressingOffset<Immediate, ShiftImmediate> {
    pub sign: Sign,
    pub value: AddressingOffsetValue<Immediate, ShiftImmediate>,
    pub mode: OffsetMode,
}

#[derive(Debug, PartialEq)]
pub enum AddressingOffsetValue<Immediate, ShiftImmediate> {
    Immediate(Immediate),
    Register(Rm),
    ScaledRegister(ShiftedRegister<ShiftImmediate>),
}

#[derive(Debug, PartialEq)]
pub enum OffsetMode {
    Offset,
    PreIndexed,
    PostIndexed,
}

#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Shift {
    LogicalShiftLeft = 0b00,
    LogicalShiftRight = 0b01,
    ArithmeticShiftRight = 0b10,
    RotateRight = 0b11,
}

#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Condition {
    /// equal
    EQ = 0b0000,

    /// not equal
    NE = 0b0001,

    /// unsigned higher or same (alias: HS)
    CS = 0b0010,

    /// unsigned lower (alias: LO)
    CC = 0b0011,

    /// negative
    MI = 0b0100,

    /// positive or zero
    PL = 0b0101,

    /// overflow
    VS = 0b0110,

    /// no overflow
    VC = 0b0111,

    //// unsigned higher
    HI = 0b1000,

    /// unsigned lower or same
    LS = 0b1001,

    /// greater or equal
    GE = 0b1010,

    /// less than
    LT = 0b1011,

    /// greater than
    GT = 0b1100,

    /// less than or equal
    LE = 0b1101,

    /// always
    AL = 0b1110,
}

pub trait RegisterIdentifier {
    fn number(&self) -> u8;
}

#[derive(Debug, PartialEq, Clone, Copy)]
pub struct Rn(pub u8);

impl RegisterIdentifier for Rn {
    fn number(&self) -> u8 {
        self.0
    }
}

#[derive(Debug, PartialEq, Clone, Copy)]
pub struct Rd(pub u8);

impl RegisterIdentifier for Rd {
    fn number(&self) -> u8 {
        self.0
    }
}

#[derive(Debug, PartialEq, Clone, Copy)]
pub struct Rm(pub u8);

impl RegisterIdentifier for Rm {
    fn number(&self) -> u8 {
        self.0
    }
}

#[derive(Debug, PartialEq, Clone, Copy)]
pub struct Rs(pub u8);

impl RegisterIdentifier for Rs {
    fn number(&self) -> u8 {
        self.0
    }
}

#[derive(Debug, PartialEq)]
pub enum ShifterOperandCode<Immediate = RotatedImm8, ShiftImm = Imm<5>> {
    Immediate(Immediate),
    ImmediateShift(ShiftedRegister<ShiftImm>),
    RegisterShift(ShiftedRegister<Rs>),
}

#[derive(Debug, PartialEq)]
pub struct ShiftedRegister<Amount, Base = Rm> {
    pub kind: Shift,
    pub amount: Amount,
    pub base: Base,
}

#[derive(Debug, PartialEq, Clone, Copy)]
pub enum MultipleAddressingMode {
    DecrementAfter,
    IncrementAfter,
    DecrementBefore,
    IncrementBefore,
}

#[derive(Debug, PartialEq)]
pub struct RegisterList {
    pub registers: [bool; 16],
}

#[derive(Debug, PartialEq)]
pub struct Imm<const N: u32>(u32);

impl<const N: u32> TryFrom<u32> for Imm<N> {
    type Error = UnencodableValueError<u32>;

    fn try_from(value: u32) -> Result<Imm<N>, Self::Error> {
        if value < (2 ^ N) {
            Ok(Imm(value))
        } else {
            Err(UnencodableValueError { value })
        }
    }
}

impl<const N: u32> Imm<N> {
    pub fn new(value: u32) -> Self {
        Self(value)
    }
    pub fn get(&self) -> u32 {
        self.0
    }
}

#[derive(Debug, PartialEq)]
pub struct RotatedImm8 {
    /// 4 bits in instruction
    rotate: u8,
    /// 8 bits in instruction
    value: u8,
}

#[derive(Debug)]
pub struct UnencodableValueError<N> {
    pub value: N,
}

impl TryFrom<u32> for RotatedImm8 {
    type Error = UnencodableValueError<u32>;

    /// tries to create a new RotatedImm8 with the specified value
    fn try_from(value: u32) -> Result<Self, Self::Error> {
        // the quickest way to find an encoding is to find the nearest encodable value <= n, and hope it equals n
        let nearest_below = RotatedImm8::nearest_below(value);

        if nearest_below.get() == value {
            Ok(nearest_below)
        } else {
            Err(UnencodableValueError { value })
        }
    }
}

impl RotatedImm8 {
    pub fn new(value: u8, rotate: u8) -> Self {
        Self { value, rotate }
    }

    /// creates a new RotatedImm8 with the nearest encodable value below or above the one specified, and also returns its offset from the specified value
    pub fn nearest_with_remainder(n: u32) -> (RotatedImm8, i32) {
        let nearest_below = RotatedImm8::nearest_below(n);
        let nearest_above = nearest_below.next();

        let nearest_below_diff = n - nearest_below.get();
        let nearest_above_diff = nearest_above.get() - n;

        if n - nearest_below.get() < nearest_above.get() - n {
            (nearest_below, nearest_below_diff as i32)
        } else {
            (nearest_above, -(nearest_above_diff as i32))
        }
    }

    /// Creates a new RotatedImm8 with the nearest encodable value below the one specified.
    /// If the specified value is encodable, it will encode that value.
    pub fn nearest_below(n: u32) -> RotatedImm8 {
        // we essentially need to choose an 8-bit window to use (window_bottom..window_top)
        // according to the ARM manual, the assembler should choose the smallest rotate possible
        // see page A5-7

        // the number of leading bits
        // capped at 24 because according to the ARM manual, if the number fits in the final 8 bits, we shouldn't rotate
        let leading = u32::min(n.leading_zeros(), 24);

        // the number of trailing zeros
        let trailing = n.trailing_zeros();

        // if there is an 8 bit window that contains all 1s, then we can use that, with the least rotate possible
        // e.g. 0 0|0 0 1 0 0 0 0 1|0 00 0 0 0 0 0 0 1
        //          ^ ^ ^ ^ ^ ^ ^ ^
        // otherwise, we use the least significant window containing the most significant bit
        // e.g. 0 0 0 0|1 0 0 0 0 0 0 0|0 0 0 0 0 0 0 1
        //              ^ ^ ^ ^ ^ ^ ^ ^
        let rotate_amount = u32::min((8 + leading) % 32, 32 - trailing);

        // encode the rotate_amount as a 4 bit value
        // we divide the rotate amount by 2 because ARM only allows rotates by even numbers. this rounds down so it will still capture the most significant 1
        // e.g. 0 0|0 1 0 0 0 0 0 0|0 0 0 0 0 0 0 0 0 0
        //          ^ ^ ^ ^ ^ ^ ^ ^
        let rotate = rotate_amount / 2;

        // get the value by rotating left (working backwards) and taking the least significant byte
        let value = n.rotate_left(rotate * 2) & 0xFF;

        RotatedImm8 {
            // must fit in 8 bits because we mod by 32 and then divide by 2
            rotate: rotate as u8,
            // must fit in 8 bits because we mod by take the least significant byte
            value: value as u8,
        }
    }

    /// calculates the next encodable value
    pub fn next(&self) -> RotatedImm8 {
        let value = self.value.wrapping_add(1);

        // if the value wraps, we need to rotate more to allow a larger value
        if value == 0 {
            let rotate = self.rotate + 1;

            // when rotated by rotate * 2, the 1 will be at the bit above the previous value
            let value = 0b01000000;

            RotatedImm8 { rotate, value }
        } else {
            RotatedImm8 {
                rotate: self.rotate,
                value,
            }
        }
    }

    pub fn value(&self) -> u8 {
        self.value
    }

    pub fn rotate(&self) -> u8 {
        self.rotate
    }

    /// gets the value
    pub fn get(&self) -> u32 {
        (self.value as u32).rotate_right(self.rotate as u32 * 2)
    }
}

#[derive(Debug, PartialEq)]
pub struct SignedImm<const N: u32>(i32);

impl<const N: u32> TryFrom<i32> for SignedImm<N> {
    type Error = UnencodableValueError<i32>;

    fn try_from(value: i32) -> Result<Self, Self::Error> {
        if (-(2 ^ (N as i32 - 1))..(2 ^ (N as i32 - 1))).contains(&value) {
            Ok(SignedImm(value))
        } else {
            Err(UnencodableValueError { value })
        }
    }
}

impl<const N: u32> SignedImm<N> {
    pub fn new(value: i32) -> Self {
        Self(value)
    }

    pub fn get(&self) -> i32 {
        self.0
    }
}

#[derive(Debug, PartialEq)]
pub enum BranchKind {
    Branch,
    BranchWithLink,
}

#[derive(Debug, PartialEq)]
pub enum MoveKind {
    Move,
    MoveNot,
}

#[derive(Debug, PartialEq)]
pub enum SetFlags {
    Set,
    DontSet,
}

#[derive(Debug, PartialEq)]
pub enum WriteBack {
    WriteBack,
    NoWriteBack,
}

#[derive(Debug, PartialEq)]
pub enum LoadStoreQuantity {
    Byte,
    Word,
}
