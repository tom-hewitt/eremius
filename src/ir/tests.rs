use crate::ir::RotatedImm8;

#[test]
fn test_rotated_imm_encoding() {
    let imm = RotatedImm8::try_from(0x0003FC00).unwrap();

    assert_eq!(imm.value(), 0xFF);
    assert_eq!(imm.rotate(), 0xB);
    assert_eq!(imm.get(), 0x0003FC00);
}

#[test]
fn test_rotated_imm_nearest_with_remainder() {
    let (imm, remainder) = RotatedImm8::nearest_with_remainder(0b11000000000000000000000000000011);

    assert_eq!(imm.value(), 0b11000000);
    assert_eq!(imm.rotate(), 4);
    assert_eq!(imm.get(), 0b11000000000000000000000000000000);
    assert_eq!(remainder, 0b11);
}

#[test]
fn test_rotated_imm_nearest_with_negative_remainder() {
    let (imm, remainder) = RotatedImm8::nearest_with_remainder(0b01111111111111111111111111111111);

    assert_eq!(imm.value(), 0b10000000);
    assert_eq!(imm.rotate(), 4);
    assert_eq!(imm.get(), 0b10000000000000000000000000000000);
    assert_eq!(remainder, -1);
}
