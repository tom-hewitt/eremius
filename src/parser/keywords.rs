// note: we use phf instead of match statements because it allows for case-insensitivity and constant lookup time
use phf::phf_map;
use unicase::UniCase;

use crate::ir::{Condition, MultipleAddressingMode};

use super::{statements::Register, Shift};

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Mnemonic {
    // Branch Instructions
    B {
        l: bool,
        condition: Condition,
    },

    // Data Processing Instructions
    ADD {
        condition: Condition,
        s: bool,
    },
    SUB {
        condition: Condition,
        s: bool,
    },
    CMP {
        condition: Condition,
    },
    MOV {
        condition: Condition,
        s: bool,
    },

    // Data Transfer Instructions
    LDR {
        condition: Condition,
    },
    STR {
        condition: Condition,
    },
    LDRB {
        condition: Condition,
    },
    STRB {
        condition: Condition,
    },
    LDM {
        condition: Condition,
        mode: MultipleAddressingMode,
    },
    STM {
        condition: Condition,
        mode: MultipleAddressingMode,
    },

    // System Calls
    SVC {
        condition: Condition,
    },

    // Pseudo Instructions
    ADR {
        condition: Condition,
        l: bool,
    },

    // Assembler Directives
    DEFW,
    DEFB,
    DEFS,
    ALIGN,
    ORIGIN,
    ENTRY,
    EQU,
}

// mnemonic and condition perfect hash map
// generated in build.rs
include!(concat!(env!("OUT_DIR"), "/codegen.rs"));

// register lookup table
pub static REGISTERS: phf::Map<UniCase<&'static str>, Register> = phf_map! {
    UniCase::ascii("R0") => Register(0),
    UniCase::ascii("R1") => Register(1),
    UniCase::ascii("R2") => Register(2),
    UniCase::ascii("R3") => Register(3),
    UniCase::ascii("R4") => Register(4),
    UniCase::ascii("R5") => Register(5),
    UniCase::ascii("R6") => Register(6),
    UniCase::ascii("R7") => Register(7),
    UniCase::ascii("R8") => Register(8),
    UniCase::ascii("R9") => Register(9),
    UniCase::ascii("R10") => Register(10),
    UniCase::ascii("R11") => Register(11),
    UniCase::ascii("R12") => Register(12),
    // stack pointer
    UniCase::ascii("SP") => Register(13),
    UniCase::ascii("R13") => Register(13),
    // link register
    UniCase::ascii("LR") => Register(14),
    UniCase::ascii("R14") => Register(14),
    // program counter
    UniCase::ascii("PC") => Register(15),
    UniCase::ascii("R15") => Register(15),
};

/// included RRX
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum ShiftName {
    LogicalShiftLeft,
    LogicalShiftRight,
    ArithmeticShiftRight,
    RotateRight,
    RotateRightExtended,
}

impl TryInto<Shift> for ShiftName {
    type Error = ();

    fn try_into(self) -> Result<Shift, Self::Error> {
        match self {
            ShiftName::LogicalShiftLeft => Ok(Shift::LogicalShiftLeft),
            ShiftName::LogicalShiftRight => Ok(Shift::LogicalShiftRight),
            ShiftName::ArithmeticShiftRight => Ok(Shift::ArithmeticShiftRight),
            ShiftName::RotateRight => Ok(Shift::RotateRight),
            ShiftName::RotateRightExtended => Err(()),
        }
    }
}

// shift lookup table
pub static SHIFT_NAMES: phf::Map<UniCase<&'static str>, ShiftName> = phf_map! {
    UniCase::ascii("LSL") => ShiftName::LogicalShiftLeft,
    UniCase::ascii("LSR") => ShiftName::LogicalShiftRight,
    UniCase::ascii("ASR") => ShiftName::ArithmeticShiftRight,
    UniCase::ascii("ROR") => ShiftName::RotateRight,
    UniCase::ascii("RRX") => ShiftName::RotateRightExtended,
};

pub static SHIFT_KINDS: phf::Map<UniCase<&'static str>, Shift> = phf_map! {
    UniCase::ascii("LSL") => Shift::LogicalShiftLeft,
    UniCase::ascii("LSR") => Shift::LogicalShiftRight,
    UniCase::ascii("ASR") => Shift::ArithmeticShiftRight,
    UniCase::ascii("ROR") => Shift::RotateRight,
};
