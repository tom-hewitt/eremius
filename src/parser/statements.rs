use std::collections::HashMap;

use crate::ir::{
    Condition, InstructionKind, LoadStoreAddressCode, Rd, Rm, Rn, Rs, ShiftedRegister,
};

#[derive(Debug, Clone, Hash, PartialEq, Eq)]
pub struct Symbol(pub String);

pub type StatementInstructionKind =
    InstructionKind<Symbol, LoadStoreAddress, ShifterOperandExpression, Expression>;

#[derive(Debug, PartialEq)]
pub enum Statement {
    Instruction { kind: StatementInstructionKind },
    PseudoInstruction { kind: PseudoInstructionKind },
    Directive { kind: DirectiveKind },
}

#[derive(Debug, PartialEq)]
pub enum PseudoInstructionKind {
    LoadRegisterConstant {
        condition: Condition,
        destination: Rd,
        value: Expression,
    },

    AddressRegister {
        condition: Condition,
        long: bool,
        destination: Register,
        label: Symbol,
    },
}

#[derive(Debug, PartialEq)]
pub enum DirectiveKind {
    Definition { kind: DefinitionKind },
    Align,
    Origin { address: Expression },
    EntryPoint,
    Constant { value: Expression },
}

#[derive(Debug, PartialEq)]
pub enum LoadStoreAddress {
    Expression(Expression),
    AddressingMode(LoadStoreAddressCode<Expression, Expression>),
}

#[derive(Debug, PartialEq)]
pub enum DefinitionKind {
    Space { size: usize, fill: Option<u8> },
    Bytes { bytes: Vec<BytesDefinition> },
    Words { words: Vec<u32> },
}

#[derive(Debug, PartialEq)]
pub enum BytesDefinition {
    Byte(u8),
    String(String),
}

impl IntoIterator for BytesDefinition {
    type Item = u8;
    type IntoIter = BytesDefinitionIter;

    fn into_iter(self) -> Self::IntoIter {
        BytesDefinitionIter {
            bytes: self,
            index: 0,
        }
    }
}

// zero alloc iterator for byte definitions
pub struct BytesDefinitionIter {
    bytes: BytesDefinition,
    index: usize,
}

impl Iterator for BytesDefinitionIter {
    type Item = u8;

    fn next(&mut self) -> Option<Self::Item> {
        match &self.bytes {
            BytesDefinition::Byte(byte) => {
                if self.index == 0 {
                    Some(*byte)
                } else {
                    None
                }
            }
            BytesDefinition::String(string) => {
                if self.index < string.as_bytes().len() {
                    Some(string.as_bytes()[self.index])
                } else {
                    None
                }
            }
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Expression {
    Number { base: u32, n: u32 },
    Character(char),
    String(String),
    Boolean(bool),
    Symbol(Symbol),
    Diadic(Box<Expression>, DiadicOperator, Box<Expression>),
}

#[derive(Debug, Clone, PartialEq)]
pub enum DiadicOperator {
    Plus,
    Minus,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Register(pub u8);

impl Into<Rd> for Register {
    fn into(self) -> Rd {
        Rd(self.0)
    }
}

impl Into<Rn> for Register {
    fn into(self) -> Rn {
        Rn(self.0)
    }
}

impl Into<Rm> for Register {
    fn into(self) -> Rm {
        Rm(self.0)
    }
}

impl Into<Rs> for Register {
    fn into(self) -> Rs {
        Rs(self.0)
    }
}

#[derive(Debug, PartialEq)]
pub enum ShifterOperandExpression {
    Immediate(Expression),
    Register(Register),
    ShiftedRegister(ShiftedRegister<ShifterOperandShiftAmount, Register>),
    RotateRightWithExtend(Register),
}

#[derive(Debug, PartialEq)]
pub enum ShifterOperandShiftAmount {
    Immediate(Expression),
    Register(Register),
}
