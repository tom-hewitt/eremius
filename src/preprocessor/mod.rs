use std::collections::HashMap;
use std::iter;

use smallvec::SmallVec;

use crate::assembler::AssemblyError;
use crate::parser::{
    DefinitionKind, DirectiveKind, Expression, Line, ParseError, Parser, PseudoInstructionKind,
    Statement, StatementInstructionKind, Symbol,
};
use crate::resolver::SymbolTable;

mod tests;

#[derive(Debug)]
pub enum PreProcessedStatement {
    Instruction { kind: StatementInstructionKind },
    PseudoInstruction { kind: PseudoInstructionKind },
    // most data definitions will be a single word, so we can increase the performance by using a small vector, which doesn't allocate until its length is greater than 4
    Data(SmallVec<[u8; 4]>),
}

#[derive(Debug)]
pub enum PreProcessError {
    ParseError(ParseError),
    OriginAddressError,
}

impl<'a> Parser<'a> {
    pub fn preprocess(self) -> Result<PreProcessResult, AssemblyError> {
        Ok(PreProcessor::new().run(self)?)
    }
}

struct PreProcessor {
    /// a pair of an address and a statement
    statements: Vec<(usize, PreProcessedStatement)>,
    symbol_table: SymbolTable<Expression>,
    entry_point: usize,
    /// maps the address of a statement to the source line that generated it
    pub source_map: HashMap<usize, usize>,
    address: usize,
    /// a queue of labels to be inserted into the symbol table at the next address
    label_queue: Vec<String>,
}

#[derive(Debug)]
pub struct PreProcessResult {
    pub statements: Vec<(usize, PreProcessedStatement)>,
    pub symbol_table: SymbolTable<Expression>,
    pub entry_point: usize,
    /// maps the address of a statement to the source line that generated it
    pub source_map: HashMap<usize, usize>,
}

impl PreProcessor {
    pub fn new() -> Self {
        Self {
            statements: Vec::new(),
            symbol_table: SymbolTable::new(),
            entry_point: 0,
            source_map: HashMap::new(),
            address: 0,
            label_queue: Vec::new(),
        }
    }

    pub fn run<'a>(
        mut self,
        lines: impl Iterator<Item = Result<Line, ParseError>> + 'a,
    ) -> Result<PreProcessResult, PreProcessError> {
        for (source_line, line) in lines.enumerate() {
            match line {
                Err(e) => return Err(PreProcessError::ParseError(e)),

                Ok(line) => {
                    // insert the label into the symbol table
                    match line.label {
                        Some(label) => self.label_queue.push(label),
                        None => (),
                    }

                    match line.statement {
                        None => (),

                        Some(statement) => match statement {
                            Statement::Instruction { kind } => self.insert_addressed_statement(
                                PreProcessedStatement::Instruction { kind },
                                4,
                                source_line,
                            ),

                            Statement::PseudoInstruction { kind } => match kind {
                                PseudoInstructionKind::AddressRegister { long, .. } => self
                                    .insert_addressed_statement(
                                        PreProcessedStatement::PseudoInstruction { kind },
                                        if long { 8 } else { 4 },
                                        source_line,
                                    ),

                                PseudoInstructionKind::LoadRegisterConstant { .. } => self
                                    .insert_addressed_statement(
                                        PreProcessedStatement::PseudoInstruction { kind },
                                        4,
                                        source_line,
                                    ),
                            },

                            // we need to apply assembler directives
                            Statement::Directive { kind } => match kind {
                                DirectiveKind::Definition { kind } => {
                                    let bytes: SmallVec<[u8; 4]> = match kind {
                                        DefinitionKind::Space { size, fill } => {
                                            iter::repeat(fill.unwrap_or(0)).take(size).collect()
                                        }
                                        DefinitionKind::Bytes { bytes } => bytes
                                            .into_iter()
                                            // use flat map with an inner iterator to avoid unnecessary allocations
                                            .flat_map(|bytes| bytes.into_iter())
                                            .collect(),
                                        DefinitionKind::Words { words } => {
                                            words.into_iter().flat_map(u32::to_be_bytes).collect()
                                        }
                                    };

                                    let size = bytes.len();

                                    self.insert_addressed_statement(
                                        PreProcessedStatement::Data(bytes),
                                        size,
                                        source_line,
                                    );
                                }

                                DirectiveKind::Align => {
                                    self.address += 4 - (self.address % 4);
                                }

                                DirectiveKind::Constant { value } => {
                                    for label in self.label_queue.drain(..) {
                                        self.symbol_table.insert(Symbol(label), value.clone());
                                    }
                                }

                                DirectiveKind::Origin { address } => {
                                    self.address =
                                        match address.backwards_resolve(&self.symbol_table) {
                                            Ok(n) => n as usize,
                                            _ => return Err(PreProcessError::OriginAddressError),
                                        };
                                }

                                DirectiveKind::EntryPoint => {
                                    self.entry_point = self.address;
                                }
                            },
                        },
                    }
                }
            }
        }

        Ok(PreProcessResult {
            statements: self.statements,
            symbol_table: self.symbol_table,
            entry_point: self.entry_point,
            source_map: self.source_map,
        })
    }

    fn insert_addressed_statement(
        &mut self,
        statement: PreProcessedStatement,
        size: usize,
        source_line: usize,
    ) {
        for label in self.label_queue.drain(..) {
            self.symbol_table.insert(
                Symbol(label),
                Expression::Number {
                    base: 10,
                    n: self.address as u32,
                },
            );
        }

        self.source_map.insert(self.address, source_line);

        self.statements.push((self.address, statement));

        self.address += size;
    }
}
