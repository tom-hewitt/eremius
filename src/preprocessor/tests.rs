use crate::ir::{CalculationKind, Condition, DataProcessingKind, InstructionKind, Rd, Rn};
use crate::parser::{
    DiadicOperator, DirectiveKind, Expression, Line, PseudoInstructionKind, Register, SetFlags,
    ShifterOperandExpression, Statement, Symbol,
};

use crate::preprocessor::PreProcessor;

#[test]
fn test_label_adrl_example() {
    let lines = [
        Ok(Line {
            label: Some("func".to_owned()),
            statement: None,
        }),
        Ok(Line {
            label: None,
            statement: Some(Statement::PseudoInstruction {
                kind: PseudoInstructionKind::LoadRegisterConstant {
                    condition: Condition::AL,
                    destination: Rd(15),
                    value: Expression::Symbol(Symbol("func".to_owned())),
                },
            }),
        }),
    ];

    let result = PreProcessor::new().run(lines.into_iter());

    insta::assert_debug_snapshot!(result);
}

#[test]
fn test_origin() {
    let lines = [
        Ok(Line {
            label: Some("func".to_owned()),
            statement: Some(Statement::Instruction {
                kind: InstructionKind::DataProcessing {
                    condition: Condition::AL,
                    kind: DataProcessingKind::Calculation {
                        kind: CalculationKind::ADD,
                        set_flags: SetFlags::DontSet,
                        destination: Rd(0),
                        source: Rn(1),
                        shifter: ShifterOperandExpression::Register(Register(2)),
                    },
                },
            }),
        }),
        Ok(Line {
            label: None,
            statement: Some(Statement::Directive {
                kind: DirectiveKind::Origin {
                    address: Expression::Diadic(
                        Box::new(Expression::Symbol(Symbol("func".to_owned()))),
                        DiadicOperator::Plus,
                        Box::new(Expression::Number { base: 10, n: 12 }),
                    ),
                },
            }),
        }),
        Ok(Line {
            label: Some("func".to_owned()),
            statement: Some(Statement::Instruction {
                kind: InstructionKind::DataProcessing {
                    condition: Condition::AL,
                    kind: DataProcessingKind::Calculation {
                        kind: CalculationKind::ADD,
                        set_flags: SetFlags::DontSet,
                        destination: Rd(0),
                        source: Rn(1),
                        shifter: ShifterOperandExpression::Register(Register(2)),
                    },
                },
            }),
        }),
    ];

    let result = PreProcessor::new().run(lines.into_iter());

    insta::assert_debug_snapshot!(result);
}
