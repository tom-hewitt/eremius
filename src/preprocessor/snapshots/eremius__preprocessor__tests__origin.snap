---
source: src/preprocessor/tests.rs
expression: result
---
Ok(
    PreProcessResult {
        statements: [
            (
                0,
                Instruction {
                    kind: DataProcessing {
                        condition: AL,
                        kind: Calculation {
                            kind: ADD,
                            set_flags: DontSet,
                            destination: Rd(
                                0,
                            ),
                            source: Rn(
                                1,
                            ),
                            shifter: Register(
                                Register(
                                    2,
                                ),
                            ),
                        },
                    },
                },
            ),
            (
                12,
                Instruction {
                    kind: DataProcessing {
                        condition: AL,
                        kind: Calculation {
                            kind: ADD,
                            set_flags: DontSet,
                            destination: Rd(
                                0,
                            ),
                            source: Rn(
                                1,
                            ),
                            shifter: Register(
                                Register(
                                    2,
                                ),
                            ),
                        },
                    },
                },
            ),
        ],
        symbol_table: SymbolTable {
            table: {
                Symbol(
                    "func",
                ): Number {
                    base: 10,
                    n: 12,
                },
            },
        },
        entry_point: 0,
        source_map: {
            12: 2,
            0: 0,
        },
    },
)
